<!DOCTYPE html>
<html>
<head>
  <!-- Required meta tags -->
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">

  <title>async, EventHandler and Exceptions - Roman Vaughan</title>

  <!-- Boostrap v4 -->
  <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0-beta/css/bootstrap.min.css" integrity="sha384-/Y6pD6FV/Vv2HJnA6t+vslU6fwYXjCFtcEpHbNJ0lyAFsXTsjBbfaDjzALeQsN6M" crossorigin="anonymous">
  <script src="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0-beta/js/bootstrap.min.js" integrity="sha384-h0AbiXch4ZDo7tp9hKZ4TsHbi047NrKGLO3SEJAg45jXxnGIfYzk4Si90RDIqNm1" crossorigin="anonymous"></script>

  <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha384-wvfXpqpZZVQGK6TAh5PVlGOfQNHSoD2xbE+QkPxCAFlNEevoEH3Sl0sibVcOQVnN" crossorigin="anonymous">

  <!-- Pygments code highlighting -->
  <link rel="stylesheet" href=" /css/pygments-vs.css">

  <link rel="stylesheet" href=" /css/uikit.css">
  <link rel="stylesheet" href=" /css/main.css">
</head>
<body>
  
  <div class="container">
    
  <h1><a href="/"><img src="/assets/images/emblem.svg" /></a> async, EventHandler and Exceptions</h1>
  <div class="content">
    <p>Hopefully this will </p>
<ul>
<li>Help lost developers who dealt with a similar situation to myself</li>
<li>Get feedback if this is a good approach to dealing with throwing exceptions in an <code>async</code> <code>EventHandler</code></li>
</ul>
<h2 id="throwing-exceptions-from-events"><a class="toclink" href="#throwing-exceptions-from-events">Throwing Exceptions From Events</a></h2>
<p>A developer would typically add an event to their class</p>
<pre class="highlight"><code class="language-C#">public class MyClass
{
    public event EventHandler OnSomeEvent;

    public void TriggerSomeEvent()
    {
        try
        {
            OnSomeEvent?.Invoke(this, new EventArgs());
        }
        catch(Exception ex)
        {
            // Handle exception things here
        }
    }
}</code></pre>


<p>A subscriber class may listen to the event using </p>
<pre class="highlight"><code class="language-C#">thatClass.OnSomeEvent += (sender, args) =&gt; Console.WriteLine(&quot;Some event happened&quot;);</code></pre>


<p>If the subscriber was to throw an exception, or it failed to handle an exception on it's own. It would be passed back to <code>TriggerSomeEvent</code> </p>
<h2 id="throwing-exceptions-from-async-events"><a class="toclink" href="#throwing-exceptions-from-async-events">Throwing Exceptions From <code>async</code> Events</a></h2>
<p>This is where things get tricky. When a listener class subscribes with an async delegate, and throws an exception. it does no get handled by <code>TriggerSomeEvent</code></p>
<pre class="highlight"><code class="language-C#">thatClass.OnSomeEvent += async (sender, args) =&gt; 
{
    await Task.Yield();
    throw new Exception();
}</code></pre>


<p><code>TriggerSomeEvent</code> will never catch the event for a couple of reasons. The first being that the above delegate is <code>async void</code>, there's no <code>Task</code> to store the exception into, and that the delegate is not awaited. </p>
<p>After exploring my my options, I considered using an <code>AsyncEventHandler</code></p>
<pre class="highlight"><code class="language-C#">public delegate Task AsyncEventHandler(object sender, EventArgs e);
// and
public delegate Task AsyncEventHandler&lt;TEventArgs&gt;(object sender, TEventArgs e);</code></pre>


<p>Compared to C#'s EventHander </p>
<pre class="highlight"><code class="language-C#">public delegate void EventHandler(object sender, EventArgs e);
// and
public delegate void EventHandler&lt;TEventArgs&gt;(object sender, TEventArgs e);</code></pre>


<p>The only difference being that AsyncEventHandler returns type <code>Task</code> instead of <code>void</code></p>
<p>Invoking an <code>AsyncEventHandler</code> is slightly more complicated</p>
<pre class="highlight"><code class="language-C#">public delegate Task AsyncEventHandler(object sender, EventArgs e);

public class MyEventClass
{
    /// &lt;summary&gt;
    /// A async event delegate
    /// &lt;/summary&gt;
    public AsyncEventHandler MyAsyncEvent;

    /// &lt;summary&gt;
    /// A typical event delegate
    /// &lt;/summary&gt;
    public EventHandler MyEvent;

    /// &lt;summary&gt;
    /// Invokes all listeners and awaits any that are async
    /// &lt;/summary&gt;
    /// &lt;returns&gt;&lt;/returns&gt;
    public async Task TriggerAsyncEvent()
    {
        try
        {
            await Task.WhenAll(GetAwaitableAsyncEvents()).ConfigureAwait(false);
        }
        catch (Exception ex)
        {
            // Exceptions are caught here
        }
    }

    /// &lt;summary&gt;
    /// Invoked from &lt;see cref=&quot;TriggerAsyncEvent&quot;/&gt; to return async delegates to await
    /// &lt;/summary&gt;
    /// &lt;returns&gt;&lt;/returns&gt;
    private IEnumerable&lt;Task&gt; GetAwaitableAsyncEvents()
    {
        foreach (var listener in MyAsyncEvent.GetInvocationList())
            if (listener.DynamicInvoke(this, new EventArgs()) is Task task)
                yield return task;
    }

    /// &lt;summary&gt;
    /// Typical event invokation 
    /// &lt;/summary&gt;
    public void TriggerEvent()
    {
        try
        {
            MyEvent(this, new EventArgs());
        }
        catch (Exception ex)
        {
            // Exceptions are caught here
        }
    }
}</code></pre>


<p>In both cases, calling <code>TriggerEventAsync()</code> or  <code>TriggerEvent()</code> will handle exceptions thrown in either synchronous or asynchronous code blocks.</p>
<p>I have written some tests with Nunit3 and .Net Core 1.1 to illustrate working examples of exceptions being correctly handled up the stack. </p>
<p><a href="https://gist.github.com/NZSmartie/2b66924cffc98047f995c7f3813046eb">Handle Exceptions from async EventHandler with Nunit tests </a></p>
<p>Feedback is appreciated!</p>
  </div>
  <hr>

  </div>
  <footer class="container text-center">
  
<button type="button" class="top btn btn-outline-dark" onclick="window.scrollTo(0, 0)">
  <i class="fa  fa-chevron-up" aria-hidden="true"></i>
  Top
</button>
<div class="contact">
    <a href="https://twitter.com/NZSmartie"><i class="fa fa-twitter" aria-hidden="true"></i><span class="sr-only">My Twitter</span></a>
    <a href="https://github.com/NZSmartie"><i class="fa fa-github" aria-hidden="true"></i><span class="sr-only">My Github</span></a>
    <a href="https://keybase.io/nzsmartie"><i class="fa fa-keybase" aria-hidden="true"></i><span class="sr-only">My Keybase</span></a>
    <a href="https://hackaday.io/NZSmartie"><i class="fa fa-hackaday" aria-hidden="true"></i><span class="sr-only">My Hack A Day I O Profile</span></a>
</div>

    <p class="meta text-right">
      Proudly built with <a href="https://getstatik.com/">Statik</a>
      <br />
      and hosted on <a href="https://help.github.com/articles/what-is-github-pages/">GitHub Pages</a>
    </p>
    

  </footer>
</body>
</html>